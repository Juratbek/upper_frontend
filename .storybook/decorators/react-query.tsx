import React from 'react';
import { QueryClientProvider } from 'react-query';
import { QueryClient } from 'react-query';

export const queryClient = new QueryClient();

export const ReactQueryDecorator = (Story) => (
  <QueryClientProvider client={queryClient}>
    <Story />
  </QueryClientProvider>
);
