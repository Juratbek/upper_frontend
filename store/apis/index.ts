export * from './article';
export * from './blog';
export * from './label';
export * from './published-article';
export * from './comment';
