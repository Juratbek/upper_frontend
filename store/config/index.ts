import apiClient from './api-client';
export * from './hooks';

export { apiClient };
