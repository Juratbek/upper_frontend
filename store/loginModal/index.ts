export * from './loginModalSelector';
export * from './loginModalSlice';
