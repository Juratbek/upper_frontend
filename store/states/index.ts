export * from './loginModal';
export * from './registerModal';
export * from './writeArticle';
export * from './auth';
export * from './commentsSidebar';
export * from './readArticle';
export * from './sidebar';
export * from './logoutModal';
export * from './tutorialsSidebar';
export * from './publishedTutorialSidebar';
export * from './theme';
