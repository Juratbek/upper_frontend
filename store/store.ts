import { configureStore } from '@reduxjs/toolkit';
import { createWrapper } from 'next-redux-wrapper';

import {
  articleApi,
  blogApi,
  commentApi,
  docsApi,
  feedbackApi,
  labelApi,
  labelRequestApi,
  notificationApi,
  publishedArticleApi,
  publishedTutorialApi,
  tutorialApi,
} from './apis';
import { apiErrorMiddleware } from './middlewares';
import authReducer from './states/auth';
import commentsSidebarReducer from './states/commentsSidebar';
import loginModalReducer from './states/loginModal';
import logoutModalReducer from './states/logoutModal';
import publishedTutorialSidebarReducer from './states/publishedTutorialSidebar';
import readArticleReducer from './states/readArticle';
import registerModalReducer from './states/registerModal';
import sidebarReducer from './states/sidebar';
import tutorialsSidebarReducer from './states/tutorialsSidebar';
import writeArticleReducer from './states/writeArticle';

export const store = configureStore({
  reducer: {
    loginModal: loginModalReducer,
    registerModal: registerModalReducer,
    writeArticle: writeArticleReducer,
    readArticle: readArticleReducer,
    auth: authReducer,
    comments: commentsSidebarReducer,
    sidebar: sidebarReducer,
    logoutModal: logoutModalReducer,
    tutorialsSidebar: tutorialsSidebarReducer,
    publishedTutorialSidebar: publishedTutorialSidebarReducer,
    [blogApi.reducerPath]: blogApi.reducer,
    [articleApi.reducerPath]: articleApi.reducer,
    [labelApi.reducerPath]: labelApi.reducer,
    [publishedArticleApi.reducerPath]: publishedArticleApi.reducer,
    [commentApi.reducerPath]: commentApi.reducer,
    [notificationApi.reducerPath]: notificationApi.reducer,
    [docsApi.reducerPath]: docsApi.reducer,
    [labelRequestApi.reducerPath]: labelRequestApi.reducer,
    [publishedTutorialApi.reducerPath]: publishedTutorialApi.reducer,
    [tutorialApi.reducerPath]: tutorialApi.reducer,
    [feedbackApi.reducerPath]: feedbackApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(
      blogApi.middleware,
      articleApi.middleware,
      labelApi.middleware,
      publishedArticleApi.middleware,
      commentApi.middleware,
      notificationApi.middleware,
      docsApi.middleware,
      labelRequestApi.middleware,
      publishedArticleApi.middleware,
      tutorialApi.middleware,
      feedbackApi.middleware,
      apiErrorMiddleware,
    ),
});

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const makeStore = () => store;

export type TDispatch = typeof store.dispatch;
export type TRootState = ReturnType<typeof store.getState>;

export const wrapper = createWrapper(makeStore);
