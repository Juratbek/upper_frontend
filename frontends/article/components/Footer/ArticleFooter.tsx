import { CommentIcon, DislikeIcon, LikeIcon, ShareIcon } from 'components/icons';
import { Clickable, Divider } from 'components/lib';
import { SharePopover } from 'components/organisms';
import { UPPER_BLUE_COLOR } from 'constants/colors';
import { useAuth, useClickOutside, useTheme } from 'hooks';
import { useRouter } from 'next/router';
import { FC, useState } from 'react';
import { useDispatch } from 'react-redux';
import { useCommentsCount } from 'store/clients/comments';
import { ReactionType, useHasReaction, useReact, useReactionCount } from 'store/clients/reaction';
import { toggleCommentsModal } from 'store/states';
import { getClassName } from 'utils';

import classes from './ArticleFooter.module.scss';

export const ArticleFooter: FC<{ sharePopoverId: string }> = ({ sharePopoverId }) => {
  const { query } = useRouter();
  const articleId = Number(query.id);
  const dispatch = useDispatch();
  const { isAuthenticated, openLoginPage } = useAuth();
  const { data: likeCountData, refetch: refetchLikeCount } = useReactionCount(
    articleId,
    ReactionType.LIKE,
  );

  const { data: likeData, refetch: refetchIsLiked } = useHasReaction({
    articleId,
    isAuthenticated,
    type: ReactionType.LIKE,
  });
  const { data: dislikeData, refetch: refetchIsDisliked } = useHasReaction({
    articleId,
    isAuthenticated,
    type: ReactionType.DISLIKE,
  });

  const refetchLikeDislikeQueries = () => {
    refetchLikeCount();
    refetchIsLiked();
    refetchIsDisliked();
  };

  const { mutate: like, isPending: isLikePending } = useReact(
    articleId,
    ReactionType.LIKE,
    refetchLikeDislikeQueries,
  );
  const { mutate: dislike, isPending: isDisLikePending } = useReact(
    articleId,
    ReactionType.DISLIKE,
    refetchLikeDislikeQueries,
  );

  const { data: commentsCountData } = useCommentsCount(articleId);
  const [isOpen, setIsOpen] = useState(false);
  const { themeColors } = useTheme();

  const commentClickHandler = (): unknown => dispatch(toggleCommentsModal());
  const shareClickHandler = (): unknown => setIsOpen((prev) => !prev);
  const closeSharePopover = (): unknown => setIsOpen(false);

  const [rootRef] = useClickOutside(closeSharePopover, `#${sharePopoverId}`);

  const likeHandler = (): void => {
    if (isAuthenticated) {
      like();
    } else {
      openLoginPage("Iltimos maqolaga layk bosish uchun ro'yxatdan o'ting");
    }
  };

  const dislikeHandler = (): void => {
    if (isAuthenticated) {
      dislike();
    } else {
      openLoginPage("Iltimos maqolaga dislayk bosish uchun ro'yxatdan o'ting");
    }
  };

  return (
    <div className={classes.root}>
      <div className={classes['reactions-container']}>
        <Clickable
          loading={isLikePending}
          disabled={likeData?.hasReaction || isLikePending}
          onClick={likeHandler}
          data-testid='like-icon'
        >
          <LikeIcon color={likeData?.hasReaction ? UPPER_BLUE_COLOR : themeColors.icon} />
        </Clickable>
        {Boolean(Number(likeCountData?.count)) && (
          <span className={getClassName(classes['like-count'], classes.count)}>
            {likeCountData!.count}
          </span>
        )}
        <Divider color='secondary' className={classes.divider} type='vertical' />
        <Clickable
          loading={isDisLikePending}
          disabled={dislikeData?.hasReaction || isDisLikePending}
          onClick={dislikeHandler}
        >
          <DislikeIcon color={dislikeData?.hasReaction ? UPPER_BLUE_COLOR : themeColors.icon} />
        </Clickable>
      </div>
      <Clickable
        onClick={commentClickHandler}
        data-testid='comment-icon'
        className={classes['comment-container']}
      >
        <CommentIcon color={themeColors.icon} />
        {Boolean(Number(commentsCountData?.count)) && (
          <span className={classes['count']}>{commentsCountData!.count}</span>
        )}
      </Clickable>
      <div className={classes['actions-container']}>
        {/* <Clickable>
          <SaveIcon />
        </Clickable> */}
        <Clickable
          className={classes.share}
          data-testid='share-icon'
          onClick={shareClickHandler}
          id={sharePopoverId}
        >
          <ShareIcon color={themeColors.icon} />
        </Clickable>
        <SharePopover ref={rootRef} isOpen={isOpen} closeSharePopover={closeSharePopover} />
      </div>
    </div>
  );
};
