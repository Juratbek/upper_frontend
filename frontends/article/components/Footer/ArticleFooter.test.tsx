import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { store } from 'store';
import { vi } from 'vitest';

import { ArticleFooter } from './ArticleFooter';

const qc = new QueryClient();

const Component = (
  <QueryClientProvider client={qc}>
    <Provider store={store}>
      <ArticleFooter sharePopoverId='share-popover-id' />
    </Provider>
  </QueryClientProvider>
);

const { toggleCommentsModalMock } = vi.hoisted(() => ({
  toggleCommentsModalMock: vi.fn().mockReturnValue({ type: 'TOGGLE_COMMENTS_MODAL' }),
}));

vi.mock('store/states', async (importOriginal) => {
  const actual = await importOriginal<typeof import('store/states')>();
  return {
    ...actual,
    toggleCommentsModal: toggleCommentsModalMock,
  };
});

describe('ArticleFooter', () => {
  it('initial render snapshot', () => {
    render(Component);
    expect(document.body).toMatchSnapshot();
  });

  it('opens share popover on share icon click', async () => {
    render(Component);
    const popover = screen.queryByRole('dialog', { hidden: true });
    expect(popover).not.toBeVisible();

    const shareIcon = screen.getByTestId('share-icon');
    userEvent.click(shareIcon);
    waitFor(() => {
      const popover = screen.getByRole('dialog');
      expect(popover).toBeVisible();
    });
  });

  it('opens comments modal on comment icon click', async () => {
    render(Component);

    const commentIcon = screen.getByTestId('comment-icon');
    userEvent.click(commentIcon);

    await waitFor(() => {
      expect(toggleCommentsModalMock).toHaveBeenCalledTimes(1);
    });
  });
});
