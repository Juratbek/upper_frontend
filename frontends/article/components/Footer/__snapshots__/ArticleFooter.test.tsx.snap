// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticleFooter > initial render snapshot 1`] = `
<body>
  <div>
    <div
      class="root"
    >
      <div
        class="reactions-container"
      >
        <button
          class="root undefined"
          data-testid="like-icon"
        >
          <svg
            fill="none"
            height="24"
            role="icon"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M7 11L11 2C14 2 14 5 14 5V9H19.66C19.9499 8.99672 20.2371 9.0565 20.5016 9.17522C20.7661 9.29393 21.0016 9.46873 21.1919 9.68751C21.3821 9.90629 21.5225 10.1638 21.6033 10.4423C21.6842 10.7207 21.7035 11.0134 21.66 11.3L20.28 20.3C20.2077 20.7769 19.9654 21.2116 19.5979 21.524C19.2304 21.8364 18.7623 22.0055 18.28 22H7M7 11V22M7 11H4C3.46957 11 2.96086 11.2107 2.58579 11.5858C2.21071 11.9609 2 12.4696 2 13V20C2 20.5304 2.21071 21.0391 2.58579 21.4142C2.96086 21.7893 3.46957 22 4 22H7"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.6"
            />
          </svg>
        </button>
        <div
          class="divider divider divider--vertical"
          role="divider"
        />
        <button
          class="root undefined"
          disabled=""
        >
          <svg
            fill="none"
            height="24"
            role="icon"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M17 13L13 22C12.2044 22 11.4413 21.6839 10.8787 21.1213C10.3161 20.5587 10 19.7956 10 19V15H4.34003C4.05012 15.0033 3.76297 14.9435 3.49846 14.8248C3.23395 14.7061 2.99842 14.5313 2.80817 14.3125C2.61793 14.0937 2.47753 13.8362 2.39669 13.5577C2.31586 13.2793 2.29652 12.9866 2.34003 12.7L3.72003 3.7C3.79235 3.22309 4.0346 2.78839 4.40212 2.47599C4.76965 2.16359 5.2377 1.99454 5.72003 2H17M17 13V2M17 13H19.67C20.236 13.01 20.7859 12.8119 21.2154 12.4432C21.645 12.0745 21.9242 11.5609 22 11V4C21.9242 3.43905 21.645 2.9255 21.2154 2.55681C20.7859 2.18813 20.236 1.98999 19.67 2H17"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.6"
            />
          </svg>
        </button>
      </div>
      <button
        class="root comment-container"
        data-testid="comment-icon"
      >
        <svg
          fill="none"
          height="24"
          role="icon"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M2 19.6392V5.96392C2 4.73662 2.99492 3.7417 4.22222 3.7417H19.7778C21.0051 3.7417 22 4.73662 22 5.96392V15.4185C22 16.6458 21.0051 17.6407 19.7778 17.6407H7.5125C6.83742 17.6407 5.77723 18.4747 5.77723 18.4747L3.18724 20.0556C3.18724 20.0556 2 20.7213 2 19.6392Z"
            stroke="black"
            stroke-width="1.4"
          />
          <path
            d="M6.5 13.0915H17.5"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <path
            d="M6.50101 9.09153L15.501 9.09149"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <div
        class="actions-container"
      >
        <button
          class="root share"
          data-testid="share-icon"
          id="share-popover-id"
        >
          <svg
            fill="none"
            height="24"
            role="icon"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M10.5959 13.3974L22.5997 2.78917M10.5959 13.3974L2.55763 8.95484C2.40182 8.86873 2.4277 8.63727 2.59868 8.5877L22.5997 2.78917M10.5959 13.3974L13.8261 21.7852C13.8902 21.9516 14.1235 21.9572 14.1956 21.7942L22.5997 2.78917"
              stroke="black"
              stroke-width="1.6"
            />
          </svg>
        </button>
        <div
          class="popover"
          role="dialog"
          style="display: none;"
        >
          <div
            class="popover-header"
          >
            <p
              class="headline"
            >
              Ulashish
            </p>
            <button
              class="root close-icon"
            >
              âœ•
            </button>
          </div>
          <div
            class="popover-body"
          >
            <ul
              class="media-icons"
            >
              <li
                class="media-icon"
              >
                <a
                  class="media-icon-btn"
                  href="https://t.me/share/url?url=http://localhost:3000/"
                  target="_blank"
                >
                  <svg
                    fill="none"
                    height="20"
                    role="icon"
                    viewBox="0 0 24 24"
                    width="20"
                  >
                    <path
                      d="M4.80626 9.87148L16.7519 5.25521C18.1945 4.6977 18.9159 4.41895 19.3657 4.78324C19.8156 5.14754 19.6929 5.91106 19.4474 7.43812L17.9078 17.0181C17.6912 18.3656 17.583 19.0394 17.0937 19.2612C16.6044 19.4829 16.0264 19.1202 14.8702 18.3949L13.3248 17.4252C11.3333 16.1758 10.3375 15.551 10.2362 14.5825C10.1348 13.614 10.9797 12.7967 12.6694 11.162L15.9697 7.96934C16.1366 7.80791 16.0223 7.52542 15.7901 7.52542C15.7406 7.52542 15.6921 7.53965 15.6505 7.56641L7.70166 12.6736C7.18785 13.0037 6.93094 13.1688 6.64783 13.2498C6.48484 13.2964 6.31657 13.3221 6.14709 13.3263C5.85271 13.3336 5.55821 13.2529 4.96922 13.0914C3.30827 12.6359 2.4778 12.4082 2.28331 11.984C2.17665 11.7513 2.16325 11.4866 2.24587 11.2444C2.39654 10.8027 3.19978 10.4923 4.80626 9.87148Z"
                      fill="white"
                    />
                  </svg>
                </a>
                <p
                  class="media-icon-name"
                >
                  Telegram
                </p>
              </li>
              <li
                class="media-icon"
              >
                <a
                  class="media-icon-btn"
                  href="https://linkedin.com/sharing/share-offsite/?url=http://localhost:3000/"
                  target="_blank"
                >
                  <svg
                    fill="none"
                    height="20"
                    role="icon"
                    viewBox="0 0 24 24"
                    width="20"
                  >
                    <path
                      d="M5.58789 7.22266C4.5918 7.22266 3.82031 6.44141 3.82031 5.51367C3.82031 4.58594 4.5918 3.81445 5.58789 3.81445C6.59375 3.81445 7.35547 4.58594 7.35547 5.51367C7.35547 6.44141 6.59375 7.22266 5.58789 7.22266ZM3.85938 19V8.36523H7.31641V19H3.85938ZM9.00586 19V8.36523H12.4727V10.2402H12.541C13.0684 8.90234 14.1621 8.14062 15.7734 8.14062C18.1074 8.14062 19.4258 9.63477 19.4258 12.1543V19H15.9688V12.8965C15.9688 11.6758 15.3926 10.9629 14.2598 10.9629C13.1562 10.9629 12.4727 11.7832 12.4727 12.9648V19H9.00586Z"
                      fill="white"
                    />
                  </svg>
                </a>
                <p
                  class="media-icon-name"
                >
                  LinkedIn
                </p>
              </li>
            </ul>
          </div>
          <div
            class="popover-footer"
          >
            <button
              class="root copy-container"
            >
              <div
                class="copy-icon"
              >
                <svg
                  fill="none"
                  height="24"
                  role="icon"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M6.00044 6.12787L6 15.75C6 18.75 6.75 19.5 9.50344 19.492L9.75 19.5L16.3716 19.5011C16.0624 20.3744 15.2293 21 14.25 21H9C6 21 4.5 19.492 4.5 16.5V8.25C4.5 7.27017 5.12631 6.43663 6.00044 6.12787ZM17.25 3C18.4926 3 19.5 4.00736 19.5 5.25V15.75C19.5 16.9926 18.4926 18 17.25 18H9.75C8.25 18 7.5 17.25 7.5 15.75V5.25C7.5 4.00736 8.50736 3 9.75 3H17.25ZM17.25 4.5H9.75C9.33579 4.5 9 4.83579 9 5.25V15.75C9 16.1642 9.33579 16.5 9.75 16.5H17.25C17.6642 16.5 18 16.1642 18 15.75V5.25C18 4.83579 17.6642 4.5 17.25 4.5Z"
                    fill="#003D80"
                    fill-opacity="1"
                  />
                </svg>
              </div>
              <p
                class="copy-text"
              >
                Nusxalash
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
